File: session_2/ruby/src/sut.rb
class Sut
  def initialize(collaborator)
    collaborator.register_client(self)

    @collaborator = collaborator
  end

  def collaborator_not_needed_method
    'OK'
  end

  def collaborator_needed_method
    @collaborator.do_something_important
  end
end

class Collaborator
  def register_client(client)
    @client = client
  end

  def do_something_important
    expensive_method
    'OK'
  end

  private

  def expensive_method
    raise 'I am very expensive'
  end
end

File: session_2/php/tests/Sut.php
<?php

class Sut
{
  public function __construct($collaborator)
  {
    $collaborator->registerClient($this);

    $this->collaborator = $collaborator;
  }

  public function collaboratorNotNeededMethod()
  {
    return 'OK';
  }

  public function collaboratorNeededMethod()
  {
    return $this->collaborator->doSomethingImportant();
  }
}

class Collaborator
{
  public function doSomethingImportant()
  {
    $this->expensiveMethod();
    return 'OK';
  }

  private function expensiveMethod()
  {
     throw new \Exception('Very expensive!!!!');
  }
}
